### PR TITLE
Fixed NRE in Thread Members Update

### PR DESCRIPTION
**Pull request by DWaffles**
Added a null check for some kind of edge case in thread members update. I haven't been able to reproduce the issue @VelvetThePanda had in Discord but this should fix it.

Also corrected some XML comments.

### CODE REVIEW

**Comment by Neuheit** [comment]
Do you have a reproducible code sample? We may still want to expose the event even if the thread is null.

File: DSharpPlus/Clients/DiscordClient.Dispatch.cs [hunk 1]
```diff
@@ -2002,6 +2002,10 @@ internal async Task OnThreadMemberUpdateEventAsync(DiscordThreadChannelMember me
         internal async Task OnThreadMembersUpdateEventAsync(DiscordGuild guild, ulong thread_id, IReadOnlyList<DiscordThreadChannelMember> addedMembers, IReadOnlyList<ulong?> removed_member_ids, int member_count)
         {
             var thread = this.InternalGetCachedThread(thread_id);
+
+            if (thread == null) // edge case
+                return;
```
[the remark this answers is not in the document]
Comment (VelvetThePanda) [reply DSharpPlus/Clients/DiscordClient.Dispatch.cs:2007]: It's been three months, have you checked the documentation yet?

[the remark this answers is not in the document]
Comment (VelvetThePanda) [reply DSharpPlus/Clients/DiscordClient.Dispatch.cs:2007]: It's been 6 months and three days, have you checked the documentation yet?

**Comment by OoLunar** [comment]
Closing as stale. Feel free to re-open if the PR is still required

**Comment by OoLunar** [comment]
Re-opening because I found this bug today. No longer stale, looking to merge.

**Comment by Erisa** [comment]
I get this error on the production instance of my bot every day or two, seemingly always lining with a user leaving or being banned. It never seems to have any negative consequences, nor do I know exactly ehat causes it.

I can never seem to reproduce it locally with a development instance of the bot, or force it on production. It just happens sometimes.

The error isn't very helpful without line numbers but:
```
[2022-05-14 20:32:49 +00:00] [Error] Socket handler suppressed an exception
System.NullReferenceException: Object reference not set to an instance of an object.
   at DSharpPlus.DiscordClient.OnThreadMembersUpdateEventAsync(DiscordGuild guild, UInt64 thread_id, IReadOnlyList`1 addedMembers, IReadOnlyList`1 removed_member_ids, Int32 member_count)
   at DSharpPlus.DiscordClient.HandleDispatchAsync(GatewayPayload payload)
   at DSharpPlus.DiscordClient.HandleSocketMessageAsync(String data)
   at DSharpPlus.DiscordClient.<InternalConnectAsync>g__SocketOnMessage|483_1(IWebSocketClient sender, SocketMessageEventArgs e)
```

**Comment by OoLunar** [comment]
I can confirm that the bug happens when a thread member gets banned

**Comment by VelvetThePanda** [comment]
Hey @DWaffles, threads are broken

**Comment by OoLunar** [comment]
Hey @DWaffles, threads are broken

**Comment by Erisa** [comment]
Ah, right, I can repro this now.

- Make a user join a thread.
- Make sure the thread isn't cached, I did this by archiving it and restarting the bot. May be eaasier causes.
- Make the user leave the server or be banned from the server.
- Boom. NRE.

**Comment by DWaffles** [comment]
Am looking into this now.

**Comment by DWaffles** [comment]
> Ah, right, I can repro this now.
> 
>     * Make a user join a thread.
> 
>     * Make sure the thread isn't cached, I did this by archiving it and restarting the bot. May be easier causes.
> 
>     * Make the user leave the server or be banned from the server.
> 
>     * Boom. NRE.

Okay. So, thank you for finding and listing the direct steps for reproduction. I highly appreciate it as I would not have done this series of actions on my own.

Discord does *not* send data on archived threads in a guild, which means that we cannot expect that thread to be in the cache. Turns out though if a user that was apart of an archived thread leaves the guild, Discord will still fire the `THREAD_MEMBERS_UPDATE` event for some reason. The fired `THREAD_MEMBERS_UPDATE` event is for the most recently created thread.

The band-aid fix is exactly as it is here, simply don't fire the event if the thread is null. As far as I'm aware, there's not an issue or oversight that causes active threads to not be in cache, so it's a decently safe fix. The alternative would be to construct a thread using *only* the ID and fire that, but I am adverse to such a fix.

**Comment by OoLunar** [comment]
> The band-aid fix is exactly as it is here, simply don't fire the event if the thread is null. As far as I'm aware, there's not an issue or oversight that causes active threads to not be in cache, so it's a decently safe fix. The alternative would be to construct a thread using only the ID and fire that, but I am adverse to such a fix.

What if we fetch the thread and cache it? Are we against using API calls like this?

**Comment by Erisa** [comment]
> > The band-aid fix is exactly as it is here, simply don't fire the event if the thread is null. As far as I'm aware, there's not an issue or oversight that causes active threads to not be in cache, so it's a decently safe fix. The alternative would be to construct a thread using only the ID and fire that, but I am adverse to such a fix.
> 
> What if we fetch the thread and cache it? Are we against using API calls like this?

Some users can be part of a great many archived threads at the time of leaving, so wouldn't you end up at risk of ratelimiting the bot with a large flurry of requests against archived threads? I had one user last night who left and caused 20+ of these NREs, presumably because they'd collected a mass of archived thread memberships during their time in the server.

**Comment by ExaInsanity** [comment]
> Are we against using API calls like this?

API calls should never be used to corroborate cache. that's just asking for 429s

**Comment by OoLunar** [comment]
> Some users can be part of a great many archived threads at the time of leaving, so wouldn't you end up at risk of ratelimiting the bot with a large flurry of requests against archived threads? I had one user last night who left and caused 20+ of these NREs, presumably because they'd collected a mass of archived thread memberships during their time in the server. 

That's a strong point that hadn't occurred to me. With that knowledge in mind, I'm advocating for an empty thread with just the id in case that's all the user needs (say, if they have data saved in an external database, indexed by the thread id)

**Comment by DWaffles** [comment]
Honestly, I'm not fond of creating yet an other skeleton object with just the thread id and the guild id. The alternative would be just not firing the event. While it's not really up to use to choose whether or not an event is useful for bot developers, I currently consider this to be abnormal Discord behavior. 

If others might find having thread member updated fired when someone leaves useful, then I don't have any particular qualms making a change.

**Comment by OoLunar** [comment]
Tested and ready to merge?

**Comment by DWaffles** [comment]
> Tested and ready to merge?

I've tested it and @Erisa has also agreed to test it just for a second confirmation.

**Review by Erisa — APPROVED** [review]
Tested and confirmed working from my pov, solves the NRE and doesn't introduce any regressions that I can spot from trying out a few scenarios.

**Pull request merged**